pkg/statedb/types.go

```diff
@@ -56,6 +56,19 @@ type Table[Obj any] interface {
 	// when anything in the table changes as more fine-grained notifications
 	// are not possible with a lower bound search.
 	LowerBound(ReadTxn, Query[Obj]) (iter Iterator[Obj], watch <-chan struct{})
+
+	// DeleteTracker creates a new delete tracker for the table
+	// starting from the given revision.
+	//
+	// WARNING: Do not create a WriteTxn against this table and then call this
+	// with a ReadTxn. This will lead to a dead lock as in order for DeleteTracker()
+	// to be usable by readers it needs to perform a write transaction on behalf
+	// of the caller.
+	//
+	// If a suitable WriteTxn is provided, then that is used to add the tracker.
+	// In that case the tracker will only become active when the transaction is
+	// committed.
+	DeleteTracker(txn ReadTxn, trackerName string) (*DeleteTracker[Obj], error)
```

Comment:
I think this is problematic.

Pedantically, saying this 'only' takes a `ReadTxn` is a lie, hence the potential for deadlock. AFAIU, the reason we would _want_ this nonetheless is the following situation:

Some module, for example the device controller, is the 'owner' of the devices table. Ideally, it's the only one being able to modify the table, all others should have read-only access. Thanks to this PR, we can now express this in hive, and depend on just the `Table[*Device]` in a consuming cell. If the consumer wants to use this `DeleteTracker`, though, and we disallow using a `ReadTxn` for the tracker, it's not possible.

But the primary point of having split read and write table access is to be able to reason more easily over how things can interact. If these types are 'lying' to us, this benefit goes away. An example of how this can go wrong:

```
Cell A has RWTable[t], Cell B has just Table[t]

Cell A opens WriteTxn(t)
Cell A invokes some API of Cell B (while holding the wtxn)
Cell B, in response, opens a ReadTxn on Table[t], and creates a DeleteTracker.
```

If I understand the issue correctly, this will deadlock. When attempting to understand what's going on, I'd not suspect implication of B, _since it cannot lock table t_ (as it holds a read-only reference).

I'm not sure how to resolve this. One idea I had was that maybe the delete tracker could work on its own table, but I'm not sure how feasible that is. Or maybe it can store its state outside of the table?

Reply:
Agreed. I'll do the more obvious thing and keep it taking the `WriteTxn`, but have this method in `Table`. It'll essentially be the only "write" that can be done without `RWTable`.